src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,328 @@
+    ConstBase,
+    /// `UnsafeCell` ICEs
+    Const,
+}
+
+/// Signalling data structure to ensure we don't recurse
+/// into the memory of other constants or statics
+struct IsStaticOrFn;
+
+impl<'rt, 'a, 'mir, 'tcx> InternVisitor<'rt, 'a, 'mir, 'tcx> {
+    fn intern(
+        &mut self,
+        ptr: Pointer,
+        mutability: Mutability,
+    ) -> EvalResult<'tcx, Option<IsStaticOrFn>> {
+        trace!(
+            "InternVisitor::intern {:?} with {:?}",
+            ptr, mutability,
+        );
+        // remove allocation
+        let tcx = self.ecx.tcx;
+        let memory = self.ecx.memory_mut();
+        let (kind, mut alloc) = match memory.alloc_map.remove(&ptr.alloc_id) {
+            Some(entry) => entry,
+            None => {
+                // if the pointer is dangling (neither in local nor global memory), we leave it
+                // to validation to error. The `delay_span_bug` ensures that we don't forget such
+                // a check in validation.
+                if tcx.alloc_map.lock().get(ptr.alloc_id).is_none() {
+                    tcx.sess.delay_span_bug(self.ecx.tcx.span, "tried to intern dangling pointer");
+                }
+                // treat dangling pointers like other statics
+                // just to stop trying to recurse into them
+                return Ok(Some(IsStaticOrFn));
+            },
+        };
+        // This match is just a canary for future changes to `MemoryKind`, which most likely need
+        // changes in this function.
+        match kind {
+            MemoryKind::Stack | MemoryKind::Vtable => {},
+        }
+        // Ensure llvm knows to only put this into immutable memory if the value is immutable either
+        // by being behind a reference or by being part of a static or const without interior
+        // mutability
+        alloc.mutability = mutability;
+        // link the alloc id to the actual allocation
+        let alloc = tcx.intern_const_alloc(alloc);
+        self.leftover_relocations.extend(alloc.relocations.iter().map(|&(_, ((), reloc))| reloc));
+        tcx.alloc_map.lock().set_alloc_id_memory(ptr.alloc_id, alloc);
+        Ok(None)
+    }
+}
+
+impl<'rt, 'a, 'mir, 'tcx>
+    ValueVisitor<'a, 'mir, 'tcx, CompileTimeInterpreter<'a, 'mir, 'tcx>>
+for
+    InternVisitor<'rt, 'a, 'mir, 'tcx>
+{
+    type V = MPlaceTy<'tcx>;
+
+    #[inline(always)]
+    fn ecx(&self) -> &CompileTimeEvalContext<'a, 'mir, 'tcx> {
+        &self.ecx
+    }
+
+    fn visit_aggregate(
+        &mut self,
+        mplace: MPlaceTy<'tcx>,
+        fields: impl Iterator<Item=EvalResult<'tcx, Self::V>>,
+    ) -> EvalResult<'tcx> {
+        if let Some(def) = mplace.layout.ty.ty_adt_def() {
+            if Some(def.did) == self.ecx.tcx.lang_items().unsafe_cell_type() {
+                // We are crossing over an `UnsafeCell`, we can mutate again
+                let old = std::mem::replace(&mut self.mutability, Mutability::Mutable);
```

Comment:
No I don't think this is right. This should be asserting that we are mutable. It is too late now to decide to become mutable if whoever initiated interning for this allocation screwed up and made it immutable. When interning an allocation you have to do this right; this traversal here happens after the allocation we are walking is already interned to figure out how to intern more things.

The code you had here before was mostly okay, all I was complaining about was the wording of the error message.^^

Reply:
Oh, this doesn't change the allocation's mutability, that one is still defined depending on `is_freeze` on the entire thing. This `replace` only exists to prevent `&mut` outside of `UnsafeCell` e.g. in `struct Foo { a: UnsafeCell<i32>, b: OtherType }`, where `OtherType` is not allowed to have any interior mutability, even if its allocation needs to be mutable.

Reply:
Oh. So I misunderstood what `mutability` is meant to do.

But how should that work when doing recursion? That would mean we need *two* `Mutability` in `ref_tracking`: the mutability of the new allocation, and the mutability of the value it contains.

Reply:
And also this means that this does *not* double-check that we don't accidentally have an `UnsafeCell` in an immutable allocation, right?

I am not convinced of the usefulness of the other mutability check, including the `ConstBase` check, as it will stop when encountering a reference to something that is already interned. So this does not actually reliably catch such problems in "unleash miri" mode. We still need a check in validity (eventually, not in this PR) that does [not do this skip here](https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/validity.rs#L405). That should at least be noted as a FIXME. But generally my feeling is that this is the wrong place to enforce "`const` is hereditarily frozen".

But if you want we can keep this check, it just should not *replace* the other check that we *can* do reliably during interning, which is making sure that no `UnsafeCell` is ever in an immutable allocation. Or do you think that is not worth a check because it is rather simply achieved with a "frozen" check?

Reply:
> That would mean we need two Mutability in ref_tracking: the mutability of the new allocation, and the mutability of the value it contains.

Oh I see, we do the "allocation" of the new interned thing *before* adding it to ref-tracking for eventual recursion. I had forgotten that, or did it change recently? Seems a bit strange to me or am I just missing something?

Reply:
> And also this means that this does not double-check that we don't accidentally have an UnsafeCell in an immutable allocation, right?

that's impossible. Either the type is `!freeze` by containing an `UnsafeCell` anywhere, or it is `freeze` and then we'll never hit this code here.

> I had forgotten that, or did it change recently? Seems a bit strange to me or am I just missing something?

No this is how I've been doing it from the beginning of this PR.

Reply:
> that's impossible. Either the type is !freeze by containing an UnsafeCell anywhere, or it is freeze and then we'll never hit this code here.

This relies on correctly calling `intern_mutability` before `intern`, right? Couldn't that computation be moved into `intern` so that the caller doesn't have to get this right? Also there should be a comment there saying why this is critical.

> No this is how I've been doing it from the beginning of this PR.

Okay then let's stick to it. :)

Reply:
> This relies on correctly calling intern_mutability before intern, right? Couldn't that computation be moved into intern so that the caller doesn't have to get this right?

I can create a wrapper that does both, but we still need to have direct calls with a fixed `Mutability` for for when we intern a vtable.

> Also there should be a comment there saying why this is critical.

I have added comments on the `intern_mutability` calls

Reply:
> I have added comments on the intern_mutability calls

But you didn't make it a new commit as far as I can see so I have no chance of reviewing the diff.

> I can create a wrapper that does both, but we still need to have direct calls with a fixed Mutability for for when we intern a vtable.

Hm I see. No it's okay, the number of parameters to `intern` would explode.